### DISCUSSION TITLE
Customize the quantum++ control gate

### DISCUSSION THREAD

**Discussion by zazabap** [softwareQinc/qpp#166] [Ideas]
If I have a rotational matrix and hope to control the rotation base on Rz(\theta), is it possible to implement the Rz(\theta) directly into the QCircuit with some possibility to customize the output?

**Comment by vsoftco** [comment]
@zazabap Currently not. However, `qpp::QCircuit` is a "light" object (it just stores gates), so you can have something like 

```c++
#include <iostream>

#include "qpp.h"

int main() {
    using namespace qpp;

    // parametrized gate
    auto param_gate = [](realT theta) { return gt.RX(theta); };

    for (idx i = 0; i < 10; ++i) {
        realT angle = 0 + 0.05 * i; // this is the variable we want to vary

        QCircuit qc{2, 2};
        qc.gate(gt.X, 0);
        // qc.gate(param_gate(angle), 0);
        // or can use directly a lambda as below
        qc.CTRL([](realT theta) { return gt.RX(theta); }(angle), 0, 1);
        qc.measure_all();

        QEngine qeng{qc};
        qeng.execute(1000);
        std::cout << qeng << "\n\n";
    }
}
```


There's virtually no overhead comparing to the cost of executing the circuit on an engine.

However, we will consider in the future allowing parametrized circuits.

**Reply by zazabap** [reply]
Thanks a lot for the example! I am trying to implement Continuous Time Quantum Walk and I consider this instance really useful in my study. I will try a bit more for the Rx with continuous time evolution. 

By the way, is there any possible new features under implementation like Basis Encoding or Amplitude Encoding?

**Reply by vsoftco** [reply]
@zazabap Please feel free to open an issue and we'll consider implementing it
